src/workerd/server/server.c++

```diff
@@ -2019,36 +2030,44 @@ public:
       }
 
       auto conn = kj::heap<Connection>(*this, kj::mv(cfBlobJson));
-      tasks.add(conn->http.listenHttp(kj::mv(stream.stream)).attach(kj::mv(conn)));
+
+      auto promise = kj::evalNow([&]() {
+        return conn->listedHttp.httpServer.listenHttp(kj::mv(stream.stream))
+          .attach(kj::mv(conn))
+          .attach(kj::addRef(*this));  // two attach()es because `this` must outlive `conn`
```

Comment:
I thought `.attach(a, b)` is always equivalent to `.attach(a).attach(b)`. But, from looking at the code it looks like that's only true for `kj::Own<T>::attach()`, and with `kj::Promise<T>::attach()` the destruction order is reversed?

Reply:
Yes, unfortunately, I didn't think hard enough when creating kj::Promise::attach() and it's probably too late to safely reverse the ordering. To avoid any confusion I use multiple attaches when the ordering matters.

Reply:
Ah, I see. I suppose we could deprecate both `attach()`es in favor of a fixed `attach2()` for a couple capnproto releases, then deprecate-and-rename back to `attach()`.